File: 7.lifecycle/src/lifecycle_cpp_pkg/src/lifecycle_manager.cpp
#include <chrono>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include "lifecycle_msgs/msg/state.hpp"
#include "lifecycle_msgs/msg/transition.hpp"
#include "lifecycle_msgs/srv/change_state.hpp"
#include "lifecycle_msgs/srv/get_state.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

class LifecycleManager : public rclcpp::Node
{
public:
  LifecycleManager() : Node("lifecycle_manager")
  {
    this->declare_parameter<std::vector<std::string>>(
      "managed_nodes",
      std::vector<std::string>{"sensor_station_1", "sensor_station_2"});

    managed_nodes_ = this->get_parameter("managed_nodes").as_string_array();

    // Sanitize node names (strip whitespace, quotes, and brackets)
    std::vector<std::string> sanitized_nodes;
    for (const auto & raw_name : managed_nodes_) {
      auto start = raw_name.find_first_not_of(" \t\n\r\'\"[]");
      if (start != std::string::npos) {
        auto end = raw_name.find_last_not_of(" \t\n\r\'\"[]");
        sanitized_nodes.push_back(raw_name.substr(start, end - start + 1));
      }
    }
    managed_nodes_ = std::move(sanitized_nodes);

    RCLCPP_INFO(
      get_logger(),
      "LifecycleManager initialized. Managing %zu nodes:",
      managed_nodes_.size());
    for (const auto & name : managed_nodes_) {
      RCLCPP_INFO(get_logger(), "  - /%s", name.c_str());
    }

    // Launch management sequence in worker thread to avoid blocking ROS executor
    manager_thread_ = std::thread(&LifecycleManager::run_orchestration, this);
  }

  ~LifecycleManager()
  {
    if (manager_thread_.joinable()) {
      manager_thread_.join();
    }
  }

private:
  void run_orchestration()
  {
    // Wait for nodes and services to become available
    RCLCPP_INFO(get_logger(), "==> Step 0: Waiting for lifecycle services for all nodes...");
    for (const auto & node_name : managed_nodes_) {
      auto client_change = this->create_client<lifecycle_msgs::srv::ChangeState>(
        "/" + node_name + "/change_state");
      auto client_get = this->create_client<lifecycle_msgs::srv::GetState>(
        "/" + node_name + "/get_state");

      while (!client_change->wait_for_service(2s)) {
        if (!rclcpp::ok()) {return;}
        RCLCPP_INFO(get_logger(), "Waiting for /%s/change_state...", node_name.c_str());
      }
      while (!client_get->wait_for_service(2s)) {
        if (!rclcpp::ok()) {return;}
        RCLCPP_INFO(get_logger(), "Waiting for /%s/get_state...", node_name.c_str());
      }
    }

    // Initial inspection
    print_current_states("Initial State");
    std::this_thread::sleep_for(2s);

    // Step 1: Configure all nodes
    RCLCPP_INFO(
      get_logger(),
      "==> Step 1: CONFIGURE all nodes (allocating publishers, timers, parameters)...");
    change_state_all(lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE, "configure");
    print_current_states("After Configure");

    RCLCPP_INFO(
      get_logger(),
      "[NOTE] Nodes are now INACTIVE. Timers are running, but LifecyclePublishers are inactive.");
    RCLCPP_INFO(get_logger(), "[NOTE] Monitor receives NO messages during this 4-second period.");
    std::this_thread::sleep_for(4s);

    // Step 2: Synchronously activate all nodes
    RCLCPP_INFO(
      get_logger(),
      "==> Step 2: SIMULTANEOUS ACTIVATION of all managed nodes!");
    change_state_all(lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE, "activate");
    print_current_states("After Simultaneous Activation");

    RCLCPP_INFO(
      get_logger(),
      "[NOTE] All nodes are now ACTIVE. Both stations are actively publishing messages simultaneously!");
    std::this_thread::sleep_for(8s);

    // Step 3: Synchronously deactivate all nodes
    RCLCPP_INFO(
      get_logger(),
      "==> Step 3: SIMULTANEOUS DEACTIVATION of all managed nodes!");
    change_state_all(lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE, "deactivate");
    print_current_states("After Simultaneous Deactivation");

    RCLCPP_INFO(
      get_logger(),
      "[NOTE] Nodes returned to INACTIVE. Message streams stop immediately while nodes remain alive.");
    std::this_thread::sleep_for(4s);

    // Step 4: Cleanup
    RCLCPP_INFO(get_logger(), "==> Step 4: CLEANUP all nodes (freeing publishers and timers)...");
    change_state_all(lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP, "cleanup");
    print_current_states("After Cleanup");
    std::this_thread::sleep_for(2s);

    // Step 5: Shutdown
    RCLCPP_INFO(get_logger(), "==> Step 5: SHUTDOWN all nodes to finalized state...");
    change_state_all(
      lifecycle_msgs::msg::Transition::TRANSITION_UNCONFIGURED_SHUTDOWN,
      "shutdown");
    print_current_states("After Shutdown");

    RCLCPP_INFO(
      get_logger(),
      "==> Coordinated multi-node lifecycle orchestration completed successfully!");
  }

  bool change_state(const std::string & node_name, uint8_t transition_id, const std::string & transition_name)
  {
    auto client = this->create_client<lifecycle_msgs::srv::ChangeState>(
      "/" + node_name + "/change_state");
    auto request = std::make_shared<lifecycle_msgs::srv::ChangeState::Request>();
    request->transition.id = transition_id;
    request->transition.label = transition_name;

    auto future = client->async_send_request(request);
    if (future.wait_for(5s) == std::future_status::ready) {
      auto result = future.get();
      if (result->success) {
        RCLCPP_INFO(
          get_logger(),
          "  [SUCCESS] Node /%s transitioned via '%s'",
          node_name.c_str(), transition_name.c_str());
        return true;
      }
    }
    RCLCPP_ERROR(
      get_logger(),
      "  [FAILED] Node /%s failed to transition via '%s'",
      node_name.c_str(), transition_name.c_str());
    return false;
  }

  void change_state_all(uint8_t transition_id, const std::string & transition_name)
  {
    for (const auto & node_name : managed_nodes_) {
      change_state(node_name, transition_id, transition_name);
    }
  }

  std::string get_state(const std::string & node_name)
  {
    auto client = this->create_client<lifecycle_msgs::srv::GetState>(
      "/" + node_name + "/get_state");
    auto request = std::make_shared<lifecycle_msgs::srv::GetState::Request>();

    auto future = client->async_send_request(request);
    if (future.wait_for(3s) == std::future_status::ready) {
      return future.get()->current_state.label;
    }
    return "UNKNOWN";
  }

  void print_current_states(const std::string & header)
  {
    RCLCPP_INFO(get_logger(), "--- State Check [%s] ---", header.c_str());
    for (const auto & node_name : managed_nodes_) {
      std::string state = get_state(node_name);
      RCLCPP_INFO(get_logger(), "    /%s is in state: [%s]", node_name.c_str(), state.c_str());
    }
  }

  std::vector<std::string> managed_nodes_;
  std::thread manager_thread_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LifecycleManager>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: 7.lifecycle/src/lifecycle_py_pkg/lifecycle_py_pkg/lifecycle_manager.py
#!/usr/bin/env python3
import threading
import time
from lifecycle_msgs.msg import Transition
from lifecycle_msgs.srv import ChangeState, GetState
import rclpy
from rclpy.node import Node


class LifecycleManager(Node):
    def __init__(self):
        super().__init__("lifecycle_manager")
        self.declare_parameter(
            "managed_nodes", ["sensor_station_1", "sensor_station_2"]
        )
        raw_nodes = (
            self.get_parameter("managed_nodes").get_parameter_value().string_array_value
        )
        self.managed_nodes = [
            n.strip(" \t\n\r'\"[]") for n in raw_nodes if n.strip(" \t\n\r'\"[]")
        ]

        self.get_logger().info(
            f"LifecycleManager initialized. Managing {len(self.managed_nodes)} nodes:"
        )
        for name in self.managed_nodes:
            self.get_logger().info(f"  - /{name}")

        self.worker_thread = threading.Thread(target=self.run_orchestration)
        self.worker_thread.daemon = True
        self.worker_thread.start()

    def run_orchestration(self):
        self.get_logger().info(
            "==> Step 0: Waiting for lifecycle services for all nodes..."
        )
        for node_name in self.managed_nodes:
            change_cli = self.create_client(ChangeState, f"/{node_name}/change_state")
            get_cli = self.create_client(GetState, f"/{node_name}/get_state")

            while not change_cli.wait_for_service(timeout_sec=2.0):
                if not rclpy.ok():
                    return
                self.get_logger().info(f"Waiting for /{node_name}/change_state...")
            while not get_cli.wait_for_service(timeout_sec=2.0):
                if not rclpy.ok():
                    return
                self.get_logger().info(f"Waiting for /{node_name}/get_state...")

        self.print_current_states("Initial State")
        time.sleep(2.0)

        # Step 1: Configure all nodes
        self.get_logger().info(
            "==> Step 1: CONFIGURE all nodes (allocating publishers, timers, parameters)..."
        )
        self.change_state_all(Transition.TRANSITION_CONFIGURE, "configure")
        self.print_current_states("After Configure")

        self.get_logger().info(
            "[NOTE] Nodes are now INACTIVE. Timers are ticking, but LifecyclePublishers are inactive."
        )
        self.get_logger().info(
            "[NOTE] Monitor receives NO messages during this 4-second period."
        )
        time.sleep(4.0)

        # Step 2: Synchronous Activation
        self.get_logger().info(
            "==> Step 2: SIMULTANEOUS ACTIVATION of all managed nodes!"
        )
        self.change_state_all(Transition.TRANSITION_ACTIVATE, "activate")
        self.print_current_states("After Simultaneous Activation")

        self.get_logger().info(
            "[NOTE] All nodes are now ACTIVE. Both stations are actively publishing messages simultaneously!"
        )
        time.sleep(8.0)

        # Step 3: Synchronous Deactivation
        self.get_logger().info(
            "==> Step 3: SIMULTANEOUS DEACTIVATION of all managed nodes!"
        )
        self.change_state_all(Transition.TRANSITION_DEACTIVATE, "deactivate")
        self.print_current_states("After Simultaneous Deactivation")

        self.get_logger().info(
            "[NOTE] Nodes returned to INACTIVE. Message streams stop immediately while nodes remain alive."
        )
        time.sleep(4.0)

        # Step 4: Cleanup
        self.get_logger().info(
            "==> Step 4: CLEANUP all nodes (freeing publishers and timers)..."
        )
        self.change_state_all(Transition.TRANSITION_CLEANUP, "cleanup")
        self.print_current_states("After Cleanup")
        time.sleep(2.0)

        # Step 5: Shutdown
        self.get_logger().info(
            "==> Step 5: SHUTDOWN all nodes to finalized state..."
        )
        self.change_state_all(
            Transition.TRANSITION_UNCONFIGURED_SHUTDOWN, "shutdown"
        )
        self.print_current_states("After Shutdown")

        self.get_logger().info(
            "==> Coordinated multi-node lifecycle orchestration completed successfully!"
        )

    def change_state(self, node_name: str, transition_id: int, transition_label: str) -> bool:
        client = self.create_client(ChangeState, f"/{node_name}/change_state")
        req = ChangeState.Request()
        req.transition.id = transition_id
        req.transition.label = transition_label

        future = client.call_async(req)
        # Wait for future with spin or loop
        start = time.time()
        while rclpy.ok() and not future.done():
            if time.time() - start > 5.0:
                self.get_logger().error(f"Timed out calling /{node_name}/change_state")
                return False
            time.sleep(0.05)

        if future.result() and future.result().success:
            self.get_logger().info(
                f"  [SUCCESS] Node /{node_name} transitioned via '{transition_label}'"
            )
            return True
        else:
            self.get_logger().error(
                f"  [FAILED] Node /{node_name} failed transition via '{transition_label}'"
            )
            return False

    def change_state_all(self, transition_id: int, transition_label: str):
        for node_name in self.managed_nodes:
            self.change_state(node_name, transition_id, transition_label)

    def get_state(self, node_name: str) -> str:
        client = self.create_client(GetState, f"/{node_name}/get_state")
        req = GetState.Request()
        future = client.call_async(req)
        start = time.time()
        while rclpy.ok() and not future.done():
            if time.time() - start > 3.0:
                return "TIMEOUT"
            time.sleep(0.05)

        if future.result():
            return future.result().current_state.label
        return "UNKNOWN"

    def print_current_states(self, header: str):
        self.get_logger().info(f"--- State Check [{header}] ---")
        for node_name in self.managed_nodes:
            state = self.get_state(node_name)
            self.get_logger().info(f"    /{node_name} is in state: [{state}]")


def main(args=None):
    rclpy.init(args=args)
    node = LifecycleManager()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
